15/umbraco-cms/tutorials/inspiration/migrating-macros/migrating-macros.md

```diff
@@ -0,0 +1,889 @@
+  "version": "0.1.0",
+  "extensions": [
+    {
+      "type": "blockEditorCustomView",
+      "alias": "my.blockEditorCustomView.ctaBlock",
+      "name": "Custom ctaBlock view",
+      "element": "/App_Plugins/CustomBlockViews/ctaBlock.js",
+      "forContentTypeAlias": "ctaBlock"
+    }
+  ]
+}
+```
+
+###### Richtext component view (/Views/Partials/richtext/Components/ctaBlock.cshtml)
+```csharp
+@inherits Umbraco.Cms.Web.Common.Views.UmbracoViewPage<Umbraco.Cms.Core.Models.Blocks.RichTextBlockItem>
+@{
+    var blockValue = Model.Content as CtaBlock;
+}
+<div class="Button">
+    <a href="https://www.youtube.com/watch?v=@blockValue.YoutubeVideoId">@blockValue.Title</a>
+</div>
+```
+
+# The core conversion
+No matter how you go about retrieving the relevant data or when, eventually you will end up with a raw string or a `RichTextEditorValue` that you need to convert. Let's have a look at a sample value.
+###### MacroValue
+```json
+{
+    "blocks": {
+        "contentData": [
+        ],
+        "settingsData": [
+        ]
+    },
+    "markup": "<p>Text before macro</p>\n<p>&nbsp;</p>\n<?UMBRACO_MACRO macroAlias=\"ctaButtonMacro\" title=\"CLICK HERE\" youtubeVideoId=\"xvFZjo5PgG0\" />\n<p>&nbsp;</p>\n<p>Text After Macro</p>"
+}
+```
+Let's break this down. The value holds json with
+- Empty block information
+- The markup with the actual RTE value and the inline macro data
+- The macro consists off
+  - The tag used as a placeholder where to render its output
+  - An alias to find the correct render/update logic
+  - Two paramaters with values entered by the user
+
+The first step in transforming the data is taking the json value and deserializing it into a `RichTextEditorValue`. This way we have a nice class to work with to store the updated data in.
+You deserialize it yourself, or you can use the `RichTextPropertyEditorHelper` to do the juob for you as it will also try to catch non json values that have not been migrated to the new format.
+
+###### Usage of RichtextPropertyEditorHelper
+```csharp
+RichTextPropertyEditorHelper.TryParseRichTextEditorValue(originalValue, _jsonSerializer, _logger, out var richTextEditorValue);
+```
+
+Next is to get all (relevant) macro tags out of the markup. One way of doing this is trough regex. Do not that the sample regex does not take into account that the order of parameters might be different from tag to tag. One way of dealing with this is to not take out the paramaters in the first match, but to move each paramater to a seperate regex that runs on the first match.
+
+###### Example regex
+```csharp
+// this regex does not take into account that the paramaters might be in a different order.
+private static readonly Regex MacroTagRegex = new(
+    @"<\?UMBRACO_MACRO\s+macroAlias=['""](?<macroAlias>.+)['""]\s+title=['""](?<title>.+)['""]\s+youtubeVideoId=['""](?<youtubeVideoId>.+)['""]\s*/>",
+    RegexOptions.IgnoreCase | RegexOptions.IgnorePatternWhitespace | RegexOptions.Multiline);
```

Comment:
Apparently there's an 'old' macro format, which starts with `<umbraco:macro` and uses a different attribute for the alias. Deploy already parses macros to convert local database integer IDs to UDIs and adds dependencies, so I'd suggest using the following (battle tested) regex instead:

```c#
new Regex(@"\<(?<tag>\??umbraco(?:_|:)macro) (?<attributes>.*?(macro)?Alias=""(?<alias>.+?)"".*?)?\/?\>", RegexOptions.Compiled | RegexOptions.IgnoreCase | RegexOptions.IgnorePatternWhitespace | RegexOptions.CultureInvariant);
```

You can also take a look at the `MacroParser` class in the Deploy v13 source code.

Reply:
Not important, it's up to the implementor to figure out the best way of determining macro values.